### ISSUE TITLE
Solver for linear system whose coefficient is defined by another nonlinear system? (tangent space evolution / cocycles)

### ISSUE THREAD

**Issue by tkf** [JuliaDiffEq/OrdinaryDiffEq.jl#250]
This would be highly useful for evolving tangent vectors using Jacobian; i.e., for evolving [cocycles](https://en.wikipedia.org/wiki/Oseledets_theorem#Cocycles).  This, in turn, can be used for calculating Lyapunov exponents.  IIRC similar method can be used for the Hessian-free optimization for fitting the parameters of the underlying nonlinear system.

This is originally discussed in: https://github.com/JuliaDynamics/DynamicalSystemsBase.jl/issues/17#issuecomment-360960450

A quick summary.  We need to solve:

```
dx/dt = f(t, x)      # "original" nonlinear system
dM/dt = J(t, x) M    # linear system
```

where `J = df/dx` is the Jacobian of the original nonlinear system.  `M` is a `length(x)` by `D` matrix with `1 <= D <= length(x)`.  The full state of the system is `u = [x, M]`.

My question was whether or not DifferentialEquations.jl can solve this type of system by exploiting the structure of the problem.  Namely, a large portion of the system is linear and nonlinear system does not depend on the linear system.  The latter point means that the linear system is almost like a time dependent system (if the nonlinear system is already solved elsewhere).

@ChrisRackauckas pointed to some relevant resources:

* Implicit-Explicit (IMEX) ODE: http://docs.juliadiffeq.org/latest/solvers/split_ode_solve.html#Implicit-Explicit-(IMEX)-ODE-1
* https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/test/linear_method_tests.jl

**Comment by ChrisRackauckas** [comment]
Yes, thank you for opening an issue for this. This seems related to control systems problems so I do want to address this. One simple way to do it would be to solve the nonlinear equation first and then solve the linear equation using the continuous solution to the nonlinear equation to get `A` and use a linear-only method on the second part. That will work, but we may need to optimize the interpolant calculations a bit to make the searching cheaper.

A more advanced way to do it would be via coupling the integrators. You can create an integrator for the nonlinear part and one for the linear part. Then you `step!` the nonlinear part far enough to pass the `dt` of the linear part and use the intermediate interpolation to get the values for `A` and use a linear solve method on that.

The first step though is building the linear-only specialized integrators, so this is a ways away but this should definitely be a near future target.

**Comment by tkf** [comment]
Thanks! You answered the question that I was just about to ask :)

OK, so in the advanced method, I guess this can be implemented by accessing the cache of the nonlinear system from `update_func` of `DiffEqArrayOperator` (or define some `AbstractDiffEqOperator` subtype if that makes sense).

Nice to know that some of the ingredients are coming.

**Comment by ChrisRackauckas** [comment]
>OK, so in the advanced method, I guess this can be implemented by accessing the cache of the nonlinear system from update_func of DiffEqArrayOperator (or define some AbstractDiffEqOperator subtype if that makes sense).

Yes, @jagot just opened an issue that highlights a lot of the possible methods.

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/249

And you can see how we're playing around with it already. We just need to update the coefficients of the matrix at certain times

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/perform_step/linear_perform_step.jl#L129-L154

and `expm` as necessary. And to make it more efficient, instead of `expm` we can `expmv!` using stuff from Expokit.jl

https://github.com/acroy/Expokit.jl#expmv

That way it doesn't actually require computing the full matrix exponential, just the action against a vector. 

Definitely work-in-progress though, but that basically shows how to do it.

**Comment by tkf** [comment]
Cool! Thanks for the pointers.  Expokit.jl is something good to know even outside this topic.

Confession: I don't think I'll be working on anytime soon (because I think I'll be working on discrete systems for now).  Initially, I just thought I'd use such feature if it is already available.  So waiting for the infrastructures to be settled totally works for me.

**Comment by jagot** [comment]
We need to adapt Expokit slightly though, we need a `exponentiator_cache` structure, since we don't want to reallocate work arrays for the different exponentiators every timestep.

**Comment by ChrisRackauckas** [comment]
>We need to adapt Expokit slightly though, we need a exponentiator_cache structure, since we don't want to reallocate work arrays for the different exponentiators every timestep.

Agreed. That can live in the `alg_cache` call.

**Comment by antoine-levitt** [comment]
Is this not simply solved by using forward diff / dual numbers? Also that way you're guaranteed to get the derivatives of the thing you're actually computing (whereas integrating the system would only get you approximate derivatives that become exact as dt goes to 0)

**Comment by ChrisRackauckas** [comment]
It's not because you still have to integrate the nonlinear system, and then the linear system is still defined by a quadrature with unknown evolution values since you need to have the Jacobian of the nonlinear system so you can easily do a standard quadrature method anyways, meaning it's probably best to not do it "two-pass" but instead solve the nonlinear system with Dual numbers and then update the linear system with the dual part simultaneously.

**Comment by antoine-levitt** [comment]
Not sure if we're talking about the same thing or not: the original system looks like M is the derivative of x wrt some parameter. So the natural thing would be to just integrate the system for x, and let forward diff take care of updating M?

**Comment by ChrisRackauckas** [comment]
>So the natural thing would be to just integrate the system for x, and let forward diff take care of updating M?

You would still need to perform a quadrature since it's the integration of a time-dependent linear system. I'm almost certain it's best to do that quadrature simultaneous with the original solution.

**Comment by antoine-levitt** [comment]
Just to be clear, I'm not talking about just forward differentiating f, but differentiating the whole integrator (mathematically, the time-t flow application), in which case you don't need to do anything specific to integrate M, it will get integrated by the autodiff (in a one-pass manner, as you say). But possibly we're talking about the same thing and I'm just confused.

**Comment by dkarrasch** [comment]
Some time ago, I asked a related question, possibly in the gitter channel. Yes, one could autodifferentiate the integrator, and we are doing that in our group. First tests seem to indicate that this yields more "noisy" results when doing this on a whole grid of initial conditions, when compared to finite-differencing the integrator (i.e., take initial conditions nearby and compute 2xdim ODE solutions and do the finite-differencing). For instance, I'm finding that positivity of the determinant of M(t) is violated once in a while for AD, which does not happen for the FD-version.

**Comment by antoine-levitt** [comment]
Huh, that's interesting, I would have thought the reverse: finite differencing can mess with the stepsize control and result in some noise. On the other hand, with AD you're essentially computing the *exact* derivative of a discrete dynamical system (approximating the continuous one), which should inherit more structural properties.

**Comment by dkarrasch** [comment]
In the FD-version, we integrate the FD stencil simultaneously, so they are all subject to the same stepsize control. One issue we don't understand so well currently is how tolerances affect the whole thing. Perhaps we should check how many steps and of which size are taken in the single AD-integration (in dimension 2), and in the simultaneous 4x2 FD-integration.

**Comment by ChrisRackauckas** [comment]
Operator overloading AD acts by tracing, so it can have trouble accurately handling conditions. What's happening could be related to how the norm being taken, maybe causing a step to be rejected, which then would make the estimate slightly inaccurate. 

Anyways, it shouldn't be necessary if you make use of [sensitivity analysis](http://docs.juliadiffeq.org/latest/analysis/sensitivity.html). You'll notice that this is a special case of sensitivity analysis where the parameter gradient that you're looking at is 0, so this is the sensitivity of the initial condition. From that definition, it's clear that the solution of the second part is just what the duals would be, but there's three problems. First of all, there's of course the accuracy issue with tracing around discontinuities. Secondly, operator overloading through a large problem has a large compilation time. ForwardDiff compilation times can be really long if the model is large enough.

But lastly, it's not like it's even faster. You can keep arguing that you can do it via dual numbers, and I totally agree. But that doesn't mean it's any good. Dual numbers still have to double the calculations, but you have less control. The first thing to ask is whether, if you were to write out the sensitivity equations, does that turn out to be faster than dual numbers? From my small tests before the answer is no, they tend to be the same. But that doesn't mean we can't do better. First of all, do the tolerances actually have to be the same? Usually not: usually the tolerances can be decreased on the sensitivities vs the actual system (sensitivities for line search in optimization can be approximate and work just fine for example), which is why programs like Sundials allow a difference here. That's one way to make this faster.

But that doesn't even address the question. The question is whether we could make the integrator faster if we knew part of it is linear. The answer is yes, we could split the method and do a Magnus on the bottom. This is especially nice if the nonlinear system is stiff, since then the Jacobian is "free", allowing the reuse of most of the calculations and then "almost exact" integrators can be employed there. This is essentially an IMEX algorithm, allowing the integrators to do a lot less calculations than the naive Dual number version. That's what I want to explore.

**Comment by antoine-levitt** [comment]
Interesting, thanks for the writeup!
